### ISSUE TITLE
metrics/lid.py: absolute epsilons make LID scale-dependent and duplicate-driven

### ISSUE THREAD

**Issue by cmvcordova** [latent-reasoning-works/manylatents#304]
`metrics/lid.py` has two absolute epsilons guarding a scale-free quantity, and on
duplicate-heavy embeddings they decide the answer rather than protect it.

```python
# manylatents/metrics/lid.py:46-48
r_k = distances[:, -1]
r_k = np.maximum(r_k, 1e-10)  # prevent division by zero (duplicate embeddings)
lid_values = -k / np.sum(np.log(distances / r_k[:, None] + 1e-10), axis=1)
```

LID is invariant to a positive rescale in exact arithmetic. It stops being
invariant the moment an absolute constant appears in it, and the symptom is not
a crash — it is a plausible-looking number.

### Two independent failures downstream

**1. Small activation scales.** A layer arriving at RMS `7.6e-16` (real: Evo2-7B
`blocks.31`) has every kNN distance below `1e-10`, so `r_k` is clamped to a
constant and the log-ratios collapse. Measured on a labelled fixture, the median
LID falls below `1.0` under the rescale while the unclamped estimator is
unchanged. Downstream this was read as a degenerate layer; the layer is fine —
its delta-norm AUROC is 0.808, essentially tied with the peak layer.

**2. Duplicate rows — the case the comment names.** The clamp does not prevent
the failure, it relocates it. A row with `j` exact duplicates among its `k`
neighbours contributes `j` terms of `log(1e-10 / r_k)`, a value fixed by the
constant rather than by the geometry, and those rows are *not* dropped. Measured
on synthetic clouds at `d=1536, k=20`:

| exact-duplicate fraction | median LID | frac < 1 | frac NaN |
|---|---:|---:|---:|
| 0% | 202.87 | 0.000 | 0.000 |
| 30% | 2.39 | 0.350 | 0.000 |
| 60% | 0.599 | 0.707 | 0.000 |

and at 60% duplicates the reported median moves **0.766 → 0.599 → 0.492** for
`EPS` of `1e-9 / 1e-12 / 1e-15`. A 36% swing in the output from a constant
nobody chose deliberately. A tight-cluster cloud *without* exact duplicates
(1e-3 jitter) gives a median of 22.2 and never drops below 1 — so this is
specifically duplicate-driven, not a general small-distance objection.

Local intrinsic dimension below 1 in a 1,536-dimensional space is not a
dimension, which makes it a usable canary: if `np.nanmedian(lid) < 1`, the
estimator is out of its domain.

### Why it went unnoticed for a while

Downstream code compensated in the opposite direction with a `lid < 10000`
validity ceiling, which only ever removed values the clamp had inflated. The two
masked each other, so neither looked wrong in isolation.

### Suggested fix

1. **RMS-normalise before any distance work** — divide by one finite nonzero
   global RMS, raise if it is zero or non-finite. Provably a no-op for
   well-conditioned inputs; verified as a bit-for-bit no-op on a duplicate-free
   cloud.
2. **Deduplicate before the kNN.** Group bit-identical rows (a `np.void` view of
   the raw bytes — no tolerance, nothing to tune), build the neighbour bank from
   the distinct points, and give every row the LID of its own point. Duplicates
   are the same point, so they must receive the same value; when a duplicate
   group spans classes, that shared value is the honest answer and the ceiling it
   places on any downstream AUROC is a fact about the embedding.
3. **Then drop both epsilons.** With duplicates removed no distance in the ratio
   can be zero, so they have nothing left to protect against — and leaving them
   in means the next duplicate-heavy array silently reintroduces this.
4. **Raise instead of returning a number** when the cloud has fewer than `k`
   distinct points.

A working implementation with tests, including an invariance test pinned at the
literal `7.6e-16` and an exact-no-op assertion on a duplicate-free cloud, is at
`cross-dogma/experiments/scripts/geometry_conditioning.py` in
`latent-reasoning-works/dogma-signal-manifold` (branch
`cesar/phase2-fix-layer-collapse`). Plain numpy/scipy/sklearn. Happy to port it
up as a PR if the approach looks right.

### Why this is worth fixing upstream rather than around

It is not hypothetical for us: on one real 105,224-row embedding set, 89,011
rows shared an exact duplicate, and the resulting LID correlated **−0.93** with
duplicate group size — group size alone outscored the LID it had contaminated.
Any consumer of this metric on protein or transcript embeddings, where near- and
exact-duplication is normal, is exposed to the same thing.

**Comment by cmvcordova** [comment]
Fix is up in #305, with one correction to this issue that is worth stating here because it makes the case stronger, not weaker.

I wrote above that the epsilon "decides the answer" on duplicate-heavy input, and quoted a sweep showing the median LID moving 0.766 → 0.599 → 0.492 for `1e-9 / 1e-12 / 1e-15`. That sweep was run with `torch.cdist`, which returns an exact zero for duplicate pairs. **Through `compute_knn`'s own path it does not reproduce**, and finding out why was the useful part.

`compute_knn` expands `||a||² + ||b||² − 2a·b`. A true zero distance between duplicate rows therefore surfaces as exactly 0 for some pairs and as a cancellation residue for others. On a 400-point cloud where every row has one exact duplicate:

| duplicate self-distance | share | what the old estimator did |
|---|---:|---|
| exactly `0.0` | 37% | `log(0/r_k + 1e-10)` — the answer is the epsilon |
| `~1.7e-07` residue | 63% | `log(3.3e-08)` — the answer is float32 round-off |

Neither group was dropped; both were reported. Median LID **0.913 in a 64-dimensional space**.

So the sharper claim is that **no choice of epsilon could have fixed this.** Whichever constant you pick, the 63% that land on round-off ignore it entirely and the 37% are governed by nothing else. That is why #305 deduplicates rather than retuning: the zero has to stop being in the neighbour list at all.

Everything else in the issue stands, including the small-scale failure and the −0.93 correlation with duplicate group size on the real embedding set. `tests/metrics/test_lid.py` pins the 37/63 split directly so the mechanism is checkable rather than asserted.
